### PR TITLE
Add OIDC middleware

### PR DESCRIPTION
**Pull request by meher745**
### This PR introduces comprehensive OpenID Connect (OIDC) support into the GoFr framework by adding:

**OIDC Discovery Metadata Fetching & Caching (discovery.go)**
Implements dynamic fetching and caching of OIDC provider metadata (including issuer, JWKS URI, and userinfo endpoint) from the standard .well-known/openid-configuration URL.

**OIDC Userinfo Middleware (oidc.go)**
Middleware to fetch user profile information from the OIDC userinfo endpoint after OAuth JWT token validation. This middleware injects the retrieved user info into the request context for downstream handlers to use.

**Dedicated Tests (oidc_test.go, discovery_test.go)**
Thorough test coverage for both userinfo middleware and discovery metadata fetching utilities, covering success scenarios, error handling, caching behavior, and edge cases.

**Documentation (docs/advanced-guide/oidc.md)**
Detailed usage guide explaining how to configure and use the new OIDC features alongside GoFr’s built-in OAuth middleware. Includes examples for discovery, middleware registration, and handler access to user info.

### Implementation Highlights

1. discovery.go:

- Fetches OIDC metadata from the provider’s discovery URL.
- Caches metadata with configurable expiration to minimize network calls.

2. oidc.go:

- Middleware extracts the validated Bearer token (via Authorization header).
- Calls the provider’s userinfo endpoint with the token and parses JSON user details.
- Injects user info into the Go standard context.Context, retrievable by the helper GetOIDCUserInfo.

**Usage Flow:**

- Fetch discovery metadata at app startup.
- Pass discovered JWKS URI and issuer to app.EnableOAuth.
- Register the OIDC userinfo middleware after OAuth.
- Access user profile info in route handlers via context helper.

### **Testing**
All new tests pass:

- oidc_test.go covers userinfo middleware success, token absence, userinfo endpoint failures, and JSON parsing errors.
- discovery_test.go covers metadata fetching success, caching, HTTP errors, JSON errors, and network failures.

Existing package tests:

- All existing and new OIDC-related tests pass locally.
- There is one unrelated failing legacy test (TestGetJwtClaims) from the GoFr core package that does NOT concern this PR and does not affect the OIDC middleware functionality.

**Thanks for considering this contribution!**

### CODE REVIEW

**Comment by meher745** [comment]
Hi, @Umang01-hash  @coolwednesday . Please review this PR. Thank you so much!

File: pkg/gofr/http/middleware/oidc.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+// Put this middleware **after** GoFr's OAuth middleware.
+func OIDCUserInfoMiddleware(userInfoEndpoint string) func(http.Handler) http.Handler {
+    return func(next http.Handler) http.Handler {
+        return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+            authHeader := r.Header.Get("Authorization")
+            if !strings.HasPrefix(authHeader, "Bearer ") {
+                http.Error(w, "Unauthorized: missing bearer token", http.StatusUnauthorized)
+                return
+            }
+            accessToken := strings.TrimPrefix(authHeader, "Bearer ")
```
Comment (ccoVeille) [inline pkg/gofr/http/middleware/oidc.go]: You are looking for strings.CutPrefix

File: pkg/gofr/http/middleware/discovery.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+func FetchOIDCMetadata(discoveryURL string) (*OIDCMetadata, error) {
+    mu.Lock()
+    defer mu.Unlock()
+
+    // Return cached metadata if still valid
+    if cachedMeta != nil && time.Now().Before(cacheExpiry) {
+        return cachedMeta, nil
+    }
+
+    resp, err := http.Get(discoveryURL)
```
Comment (ccoVeille) [inline pkg/gofr/http/middleware/discovery.go]: And http.Get without passing a contexf doesn't sound a good idea

File: pkg/gofr/http/middleware/discovery.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    var meta OIDCMetadata
+    if err := json.NewDecoder(resp.Body).Decode(&meta); err != nil {
+        return nil, fmt.Errorf("failed to decode OIDC discovery JSON: %w", err)
+    }
+
+    // Cache the fetched metadata
+    cachedMeta = &meta
+    cacheExpiry = time.Now().Add(cacheDuration)
```
Comment (ccoVeille) [inline pkg/gofr/http/middleware/discovery.go]: I have a problem here. The cache is absolute and at package level.

Which means that calling the method with an URL, then an another will return the information of the first one.

Also the mutex is global

I would create a struct instead

type whatever struct {
  url string
  mu mutex.RW
  cacheMeta
  cacheExpiry
}

then I would create a struct by passing the URL, and call a method on this struct to retrieve the data

**Comment by meher745** [comment]
Hi @ccoVeille @Umang01-hash @coolwednesday . I've updated the files according to the recommendations. Please review.

File: pkg/gofr/version/version.go [hunk 4]
```diff
@@ -1,3 +1,3 @@
 package version
 
-const Framework = "dev"
+const Framework = "v1.43.0"
```
Comment (ccoVeille) [inline pkg/gofr/version/version.go]: Why this change?

Comment (meher745) [reply pkg/gofr/version/version.go]: Hi, this change was not intentional. It seems it got included in my PR because I accidentally staged all changes using git add . while preparing my commit. Please let me know if you’d like me to remove it. @ccoVeille

**Comment by meher745** [comment]
@ccoVeille , kindly merge the pull request. Thanks!

**Comment by ccoVeille** [comment]
I'm an active code reviewer of gofr. But I'm not a maintainer. I cannot merge.

Let's wait for a maintainer feedback

@coolwednesday @Umang01-hash

**Comment by akshat-kumar-singhal** [comment]
@meher745 There's a lot of redundancy in your code right now - consider using the existing Auth Middleware which accepts AuthProvider. 
cc @Umang01-hash @coolwednesday

File: pkg/gofr/version/version.go (hunk 4 above)
Comment (Umang01-hash) [reply pkg/gofr/version/version.go]: @meher745 Please revert back this change.

File: pkg/gofr/http/middleware/oidc.go [hunk 5]
```diff
@@ -0,0 +1,76 @@
+// File: pkg/gofr/http/middleware/oidc.go
+
+package middleware
```
Comment (Umang01-hash) [inline pkg/gofr/http/middleware/oidc.go:1]: To integrate OIDC with the existing `AuthProvider` based **AuthMiddleware**, you can implement a new **OIDCAuthProvider** struct in `oidc.go` that satisfies the **AuthProvider** interface. This allows you to reuse the unified authentication flow for OIDC, just like Basic, API key, and OAuth. Simply extract  (& validate if needed) the Bearer token in **ExtractAuthHeader**, and return user info or claims as needed. Then, pass your provider to AuthMiddleware for consistent middleware usage.

**Review by Umang01-hash — CHANGES_REQUESTED** [review]
@meher745 Here are a few review suggestions for your PR:

- Instead of using dynamic errors (fmt.Errorf()) please use predefined errors.
- For the documentation i see we can improve on how we are portraying this feature to user, the code included should be well formatted, include proper spaces and comments. Remove repeteted things and big technical jagrons let's try to keep it simple so that evry user can understand it.
- Please resolve all the linter and code quality errors in your code.

**Comment by Umang01-hash** [comment]
@meher745 Are you still working in this PR?

**Comment by meher745** [comment]
> @meher745 Are you still working in this PR?

Yes I am. I am almost done with the desired changes.

**Comment by meher745** [comment]
Hi, @Umang01-hash @coolwednesday please let me know if the recent commits I made are fine. I'm kind of confused since I had to remove go.work.sum and go.sum, and build them again due to some import conflicts.

**Comment by meher745** [comment]
@Umang01-hash @coolwednesday Please review, thanks.

**Comment by meher745** [comment]
@ccoVeille @Umang01-hash @coolwednesday . Kindly review. Thanking you in advance!

File: go.mod
Comment (ccoVeille) [inline go.mod:1]: I don't get the version downgrade here

File: pkg/gofr/http/middleware/discovery.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	cacheExpiry   time.Time
+}
+
+// NewDiscoveryCache creates a new per-URL OIDC discovery cache.
+func NewDiscoveryCache(url string, cacheDuration time.Duration) *DiscoveryCache {
+	return &DiscoveryCache{
+		url:           url,
+		cacheDuration: cacheDuration,
+	}
+}
```
Comment (ccoVeille) [inline pkg/gofr/http/middleware/discovery.go]: url variable name would shadow url package

File: pkg/gofr/http/middleware/discovery.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	defer dc.mu.Unlock()
+
+	// Return cached metadata if still valid
+	if dc.cachedMeta != nil && time.Now().Before(dc.cacheExpiry) {
+		return dc.cachedMeta, nil
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, dc.url, http.NoBody)
+	if err != nil {
+		return nil, ErrFailedCreateDiscoveryRequest
```
Comment (ccoVeille) [inline pkg/gofr/http/middleware/discovery.go]: Here and everywhere

I would use errors.Join or fmt.Errorf("%w: %w") to keep the err in the stack

File: pkg/gofr/http/middleware/discovery.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, ErrFailedFetchDiscoveryMetadata
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, ErrBadDiscoveryStatus
```
Comment (ccoVeille) [inline pkg/gofr/http/middleware/discovery.go]: Returning the status code in the error is always interesting


```suggestion
		return nil, fmt.Errorf("%w: unexpected status code: %d", ErrBadDiscoveryStatus, resp.StatusCode)
```

File: pkg/gofr/http/middleware/oidc.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, ErrUserInfoFetch
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, ErrUserInfoBadStatus
```
Comment (ccoVeille) [inline pkg/gofr/http/middleware/oidc.go]: Same here about returning the status code

File: pkg/gofr/http/middleware/oidc.go [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+	if !ok {
+		return nil, ErrMissingToken
+	}
+
+	token = strings.TrimSpace(token)
+	if token == "" {
+		return nil, ErrEmptyToken
+	}
+
+	req, err := http.NewRequestWithContext(r.Context(), http.MethodGet, p.UserInfoEndpoint, http.NoBody)
```
Comment (akshat-kumar-singhal) [inline pkg/gofr/http/middleware/oidc.go:44]: Consider injecting the HTTP client instead of creating a new one. Also, use existing constant for 'Authorization'.

File: pkg/gofr/http/middleware/oidc.go [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+	token, ok := strings.CutPrefix(authHeader, "Bearer ")
+	if !ok {
+		return nil, ErrMissingToken
+	}
+
+	token = strings.TrimSpace(token)
+	if token == "" {
+		return nil, ErrEmptyToken
+	}
```
Comment (akshat-kumar-singhal) [inline pkg/gofr/http/middleware/oidc.go]: this is present in "getAuthHeaderFromRequest()" auth.go

File: pkg/gofr/http/middleware/oidc.go [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+// Predefined errors for consistent error handling.
+var (
+	ErrMissingToken      = errors.New("missing bearer token")
+	ErrEmptyToken        = errors.New("empty bearer token")
+	ErrCreateRequest     = errors.New("failed to create userinfo request")
+	ErrUserInfoFetch     = errors.New("failed to fetch userinfo")
+	ErrUserInfoBadStatus = errors.New("userinfo endpoint returned error status")
+	ErrUserInfoJSON      = errors.New("failed to parse userinfo response")
+)
```
Comment (akshat-kumar-singhal) [inline pkg/gofr/http/middleware/oidc.go]: avoid duplicating errors - some of these are already present

**Comment by Umang01-hash** [comment]
Hey @meher745 !

There are a few review comments on your PR that still need to be addressed. Just checking in — are you currently working on this? If so, let’s aim to wrap it up soon, as it’s been pending for a while and we’re quite close to completing it.

If you need any help or clarification, feel free to reach out — happy to support however I can.

Looking forward to your update!

**Comment by meher745** [comment]
Hi @Umang01-hash . I have gove through the suggestions and started working on them. I apologise for the delay since I had my semester exams going on. Will get back to you with updated pr soon! Thanks in advance.

**Comment by Umang01-hash** [comment]
Hi @meher745,

Thanks for the update — no worries at all! Looking forward to your updated PR.

**Comment by Umang01-hash** [comment]
@meher745 Any updated on the requested changes? The PR has been open for quite a while, and we’re hoping to close it soon. If you need any help implementing the changes, feel free to reach out. Let’s try to wrap this up together!

File: pkg/gofr/http/middleware/oidc.go (hunk 11 above)
Comment (meher745) [reply pkg/gofr/http/middleware/oidc.go]: I have to convert all errors to ErrorHTTP format for using that, to avoid that I did not use the preexisting function. If it is preferred to use the function, I will make the necessary changes. Let me know! @akshat-kumar-singhal

**Comment by meher745** [comment]
Hi @Umang01-hash @akshat-kumar-singhal . I apologize for the delay, please review the changes. Thanking you in advance!

**Comment by Umang01-hash** [comment]
@meher745 There is a merge conflict in your PR please resolve it.

**Comment by meher745** [comment]
@Umang01-hash I've resolved the conflict, please check.

File: pkg/gofr/http/middleware/oidc.go (hunk 10 above)
Comment (akshat-kumar-singhal) [reply pkg/gofr/http/middleware/oidc.go:44]: @meher745 any thoughts on this?

**Comment by Umang01-hash** [comment]
Screenshot 2025-10-29 at 11 08 22 AM


@meher745 Your PR is failing the **following tests as well as the code quality steps**. Please fix them.

**Comment by Umang01-hash** [comment]
@meher745 Are you still active on this PR??

**Comment by meher745** [comment]
Hi @Umang01-hash I will be making an updated PR by the coming Saturday. Thank you for your patience!

**Comment by meher745** [comment]
Hi @Umang01-hash, is it fine if I take a week ? My end semester exams are going on, and there seem to be some vital security checks which I don't want to rush. If it is not feasible, I will make the changes within a day. Please consider it since I have been working on this issue for the past 3 months. Thank you in advance!

**Comment by Umang01-hash** [comment]
> Hi @Umang01-hash, is it fine if I take a week ? My end semester exams are going on, and there seem to be some vital security checks which I don't want to rush. If it is not feasible, I will make the changes within a day. Please consider it since I have been working on this issue for the past 3 months. Thank you in advance!

@meher745 Sure you can take your time. Let's try we will close this in this month for sure. If you need any help from my side feel free to let me know.

**Comment by coolwednesday** [comment]
Hi @meher, as there hasn’t been any recent activity on this PR, we’ll go ahead and close it for now. I’d like to take some time to step back and analyze how adding OIDC can be most beneficial for us. If you’d like to pick this up again in the future, please let me know in the issue ticket.
